prql-compiler/src/parser/mod.rs

```diff
@@ -54,7 +54,13 @@ pub fn parse(source: &str) -> Result<Vec<Stmt>> {
 }
 
 fn convert_lexer_error(source: &str, e: Cheap<char>) -> Error {
-    let found = source[e.span()].to_string();
+    // TODO: is there a neater way of taking a span? We want to take it based on
+    // the chars, not the bytes, so can't just index into the str.
+    let found = source
+        .chars()
+        .skip(e.span().start)
+        .take(e.span().end() - e.span().start)
+        .collect();
```

Comment:
I believed that indexing str was indexing chars, not bytes! Huh, this is unexpected.

```
echo 'fn main() { let a: &str = &"😙"[1..2]; println!("{a}"); }' > a.rs && rustc a.rs && ./a
thread 'main' panicked at 'byte index 1 is not a char boundary; it is inside '😙' (bytes 0..4) of `😙`', a.rs:1:28
```

Well, after a quick look around, what you wrote seems like the best approach. Sadly.

Reply:
Using char indices instead of bytes for the default impl of `From<&str> for Stream` was an API misstep that has been corrected in `zero-copy`.